src/server/terrain/chunkgen/StalagmiteGenerator.zig

```diff
@@ -0,0 +1,155 @@
+				while(y3 <= yMax) : (y3 += 1) {
+					var z3: i32 = zMin;
+					while(z3 <= zMax) : (z3 += 1) {
+						const dist = distSqr(@as(f32, @floatFromInt(x3)) - x2, @as(f32, @floatFromInt(y3)) - y2, @as(f32, @floatFromInt(z3)) - z2);
+						if(dist < size*size) {
+							if(x3 >= 0 and x3 < chunk.super.width and y3 >= 0 and y3 < chunk.super.width and z3 >= 0 and z3 < chunk.super.width) {
+								const block: main.blocks.Block = chunk.getBlock(x3, y3, z3);
+								if(block.typ == 0 or block.degradable() or block.blockClass() == .fluid) {
+									chunk.updateBlockInGeneration(x3, y3, z3, .{.typ = stalagmite, .data = 0}); // TODO: Use natural standard.
+								}
+							}
+						}
+					}
+				}
+			}
+			if(size > 2) size = 2;
+			j += size/2; // Make sure there are no stalagmite bits floating in the air.
+			if(size < 0.5) break; // Also preventing floating stalagmite bits.
+		}
+	}
+}
+
+fn considerCoordinates(x: i32, y: i32, z: i32, chunk: *main.chunk.ServerChunk, caveMap: CaveMap.CaveMapView, biomeMap: CaveBiomeMap.CaveBiomeMapView, seed: *u64) void {
+	const oldSeed = seed.*;
+	const stalagmiteSpawns = biomeMap.getBiomeAndSeed(x +% main.chunk.chunkSize/2 -% chunk.super.pos.wx, y +% main.chunk.chunkSize/2 -% chunk.super.pos.wy, z +% main.chunk.chunkSize/2 -% chunk.super.pos.wz, true, seed).stalagmites;
+	random.scrambleSeed(seed);
+	var differendColors: u32 = 1;
+	if(random.nextInt(u1, seed) != 0) {
+		// ¹⁄₄ Chance that a cave has multiple stalagmites.
+		while(random.nextInt(u1, seed) != 0 and differendColors < 32) {
+			differendColors += 1; // Exponentially diminishing chance to have more differend stalagmites per cavern.
+		}
+	}
+	
+	// Spawn the stalagmites using the old position specific seed:
+	seed.* = oldSeed;
+	for(0..stalagmiteSpawns) |_| {
+		// Choose some in world coordinates to start generating:
+		const worldX = x + random.nextIntBounded(u31, seed, main.chunk.chunkSize);
+		const worldY = y + random.nextIntBounded(u31, seed, main.chunk.chunkSize);
+		const worldZ = z + random.nextIntBounded(u31, seed, main.chunk.chunkSize);
+		const relX = worldX -% chunk.super.pos.wx;
+		const relY = worldY -% chunk.super.pos.wy;
+		const relZ = worldZ -% chunk.super.pos.wz;
+		if(caveMap.isSolid(relX, relY, relZ)) { // Only start stalagmite in solid blocks
+			// Only start stalagmite when they are close to the surface (±SURFACE_DIST blocks)
+			if(
+				(worldX - x >= surfaceDist and !caveMap.isSolid(relX - surfaceDist, relY, relZ))
+				or (worldX - x < main.chunk.chunkSize - surfaceDist and !caveMap.isSolid(relX + surfaceDist, relY, relZ))
+				or (worldY - y >= surfaceDist and !caveMap.isSolid(relX, relY - surfaceDist, relZ))
+				or (worldY - y < main.chunk.chunkSize - surfaceDist and !caveMap.isSolid(relX, relY + surfaceDist, relZ))
+				or (worldZ - z >= surfaceDist and !caveMap.isSolid(relX, relY, relZ - surfaceDist))
+				or (worldZ - z < main.chunk.chunkSize - surfaceDist and !caveMap.isSolid(relX, relY, relZ + surfaceDist))
```

Comment:
Since stalagmites only grow upwards or downwards, maybe we only need to check the up and down direction.